pkg/proxy/proxy.go

```diff
@@ -282,12 +284,85 @@ func (r *Redirect) getDestinationInfo(dstIPPort string) EndpointInfo {
 	return info
 }
 
+const identityKey int = 0
+
+func newIdentityContext(ctx context.Context, id int) context.Context {
+	return context.WithValue(ctx, identityKey, id)
+}
+
+func identityFromContext(ctx context.Context) (int, bool) {
+	val, ok := ctx.Value(identityKey).(int)
+	return val, ok
+}
+
+func setSocketMark(c net.Conn, mark int) {
+	if tc, ok := c.(*net.TCPConn); ok {
+		if f, err := tc.File(); err == nil {
+			defer f.Close()
+			fd := int(f.Fd())
+
+			log.Debugf("Setting identity %d on socket %+v", mark, tc)
+			err := syscall.SetsockoptInt(fd, syscall.SOL_SOCKET, syscall.SO_MARK, mark)
+			if err != nil {
+				log.Debugf("Unable to set SO_MARK=%d socket option: %s", mark, err)
+			}
+		}
+	}
+}
+
+func getSocketMark(c net.Conn) int {
+	if tc, ok := c.(*net.TCPConn); ok {
+		if f, err := tc.File(); err == nil {
+			defer f.Close()
+			fd := int(f.Fd())
+
+			val, err := syscall.GetsockoptInt(fd, syscall.SOL_SOCKET, syscall.SO_MARK)
+			if err != nil {
+				log.Debugf("Unable to get SO_MARK socket option: %s", err)
+				return 0
+			}
+
+			return val
+		}
+	}
+
+	return 0
+}
+
```

Comment:
Currently this has no callers, are there planned uses? Not opposing keeping this even if not as of now.